### ISSUE TITLE
Two relations to same entity throwing exception

### ISSUE THREAD

**Issue by ghost** [dotnet/efcore#22905]
In some cases it seams like two relations to the same entity throw an exception (both testen with EF 3.1 & 5.0).

Am I doing something wrong here or is this an (known) issue of EF?

[Program.txt](https://github.com/dotnet/EntityFramework.Docs/files/5330501/Program.txt)

**Comment by ajcvickers** [comment]
Code and exception:

```C#
internal class Program
{
    private static void Main()
    {
        var context = new TestDbContext();
        context.Database.EnsureDeleted();
        context.Database.EnsureCreated();
        
        var user = context.Users.FirstOrDefault(p => 1 == 2);
    }
}

public class TestDbContext : DbContext
{
     private static readonly ILoggerFactory
         Logger = LoggerFactory.Create(x => x.AddConsole());//.SetMinimumLevel(LogLevel.Debug));

    public DbSet<User> Users { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder
            .UseLoggerFactory(Logger)
            .EnableSensitiveDataLogging()
            .UseSqlServer(Your.ConnectionString);
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<User>(u => { u.ToTable("User"); });

        modelBuilder.Entity<UserProperty>(up =>
        {
            up.ToTable("UserProperty");

            up.HasKey(p => new {p.Id, p.Name, p.ServerOnly});

            up.HasOne(p => p.User)
                .WithMany(p => p.Properties)
                .HasForeignKey(p => p.Id);
        });
    }
}

public class UserProperty : ChangeableBase
{
    public Guid Id { get; set; }
    public string Name { get; set; }
    public bool ServerOnly { get; set; }
    public string Value { get; set; }

    public virtual User User { get; set; }
}

public class User : ChangeableBase
{
    public Guid Id { get; set; }
    public virtual ICollection<UserProperty> Properties { get; set; } = new List<UserProperty>();
}

public abstract class CreateableBase
{
    public Guid? UserIdCreated { get; set; }

    [ForeignKey("UserIdCreated")] public virtual User UserCreated { get; set; }
}

public abstract class ChangeableBase : CreateableBase
{
    public Guid? UserIdChanged { get; set; }

    [ForeignKey("UserIdChanged")] public virtual User UserChanged { get; set; }
}
```

Exception:
```
warn: Microsoft.EntityFrameworkCore.Model[10612]
      Navigations 'User.UserChanged' and 'User.UserCreated' were separated into two relationships as the [ForeignKey] attribute was specified on navigations on both sides.
Unhandled exception. System.InvalidOperationException: Unable to determine the relationship represented by navigation 'UserProperty.UserChanged' of type 'User'. Either manually configure the relationship, or ignore this property using the '[NotMapped]' attribute or by using 'EntityTypeBuilder.Ignore' in 'OnModelCreating'.
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.ValidatePropertyMapping(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.SqlServer.Internal.SqlServerModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.ValidatingConvention.ProcessModelFinalized(IModel model)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelFinalized(IModel model)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelFinalized(IModel model)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.Model.FinalizeModel()
   at Microsoft.EntityFrameworkCore.ModelBuilder.FinalizeModel()
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_Dependencies()
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureDeleted()
   at Program.Main() in /home/ajcvickers/AllTogetherNow/Daily/Daily.cs:line 19
```

**Comment by AndriySvyryd** [comment]
The warning is by design. The exception is another case of https://github.com/dotnet/efcore/issues/18388

**Comment by tsproesser** [comment]
Sorry, I don't get it. Does this mean it is a known limitation? Or is something wrong with my markup?

Can I hope for a fix in future?

**Comment by AndriySvyryd** [comment]
It's a know bug, it will be fixed in 6.0.

As a workaround configure the relationships using Fluent API:
```C#
        modelBuilder.Entity<UserProperty>(up =>
        {
            up.HasOne(p => p.UserChanged)
                .WithMany();
            up.HasOne(p => p.UserCreated)
                .WithMany();
        });
```

**Comment by tsproesser** [comment]
Ah, ok! Ty very much! :) Want to go with a baseclass, but that's a workaround where it is needed.
